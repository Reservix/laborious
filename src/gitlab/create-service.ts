import got, { GotJSONOptions, GotError } from 'got';
import { URL } from 'url';

import * as pathnames from './pathnames';
import { GitlabResponse } from './responses';
import { GitlabProjectID } from './types';

export type GitlabApiConfig = {
  baseUrl: string;
  token: string;
  userAgent: string;
};

/**
 * `got.create` is currently not part of the typings and adding them failed
 * (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/31385). So we just
 * code "around" this by just using `got`.
 */
export const create = ({ baseUrl, token, userAgent }: GitlabApiConfig) => {
  const apiURL = new URL(baseUrl);
  apiURL.pathname = pathnames.API_PREFIX;

  const config: GotJSONOptions = {
    json: true,
    headers: {
      'private-token': token,
      accept: 'application/json',
      'user-agent': userAgent,
    },
  };

  return async <T>(
    method: 'get' | 'post' | 'put' | 'patch' | 'head' | 'delete',
    url: string,
    options: Pick<GotJSONOptions, 'body' | 'query'> = {}
  ) => {
    try {
      const res = await got[method](`${apiURL}${url}`, {
        ...config,
        ...options,
      });
      return res.body as T;
    } catch (e) {
      const body = (e as GotError).response.body;
      const message: string[] = Array.isArray(body.message)
        ? body.message
        : [body.message];

      const err = new Error(message.map(s => `  ${s}`).join('\n'));
      err.name = 'GitlabRequestError';
      throw err;
    }
  };
};

export const createGitlabService = (config: GitlabApiConfig) => {
  const api = create(config);

  // Version
  // ---------------
  function version() {
    return api<GitlabResponse.Version>('get', pathnames.version());
  }

  // Project
  // ---------------
  function project(pid: GitlabProjectID) {
    return api<GitlabResponse.Project>('get', pathnames.project(pid));
  }

  project.createMergeRequest = (options: {
    id: GitlabProjectID;
    source_branch: string;
    target_branch: string;
    title: string;
    assignee_id?: number;
    description?: string;
    target_project_id?: string;
    labels?: string;
    milestone_id?: string;
    remove_source_branch?: boolean;
    allow_collaboration?: boolean;
    allow_maintainer_to_push?: boolean;
    squash?: boolean;
  }) =>
    api<GitlabResponse.MergeRequest>(
      'post',
      pathnames.mergeRequests(options.id),
      { body: options }
    );

  project.listMergeRequests = (
    pid: GitlabProjectID,
    options: Partial<{ state: 'opened' | 'closed' | 'locked' | 'merged' }> = {
      state: 'opened',
    }
  ) =>
    api<GitlabResponse.MergeRequest[]>('get', pathnames.mergeRequests(pid), {
      query: options,
    });

  project.listBranches = (pid: GitlabProjectID) =>
    api<GitlabResponse.Branch[]>('get', pathnames.branches(pid));

  return {
    version,
    project,
  };
};
