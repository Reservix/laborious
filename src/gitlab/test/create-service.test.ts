import got from 'got';
import { createGitlabService } from '../create-service';

jest.mock('got', () => {
  const mock = ['get', 'post', 'put', 'patch', 'head', 'delete'].reduce(
    (o, method) => {
      o[method] = jest.fn().mockResolvedValue({ data: `<mocked ${method}>` });
      return o;
    },
    {}
  );
  return {
    ...mock,
  };
});

let service: ReturnType<typeof createGitlabService>;

beforeEach(() => {
  (got.get as jest.Mock).mockClear();
  (got.post as jest.Mock).mockClear();

  service = createGitlabService({
    baseUrl: 'https://gitlab.example.com',
    token: 'T0kn',
    userAgent: 'Testing laborious',
  });
});

test('smoke test', () => {
  expect(service).toEqual(expect.any(Object));
});

test('get version', async () => {
  await service.version();
  expect(got.get).toHaveBeenCalled();
  expect(got.get).toMatchLastCallSnapshot();
});

test('get a project', async () => {
  await service.project('ids can also be strings');
  expect(got.get).toHaveBeenCalled();
  expect(got.get).toMatchLastCallSnapshot();
});

test('create a merge request', async () => {
  await service.project.createMergeRequest({
    id: 12346,
    source_branch: 'feature/123',
    target_branch: 'master',
    title: 'this is a merge request',
  });
  expect(got.post).toHaveBeenCalled();
  expect(got.post).toMatchLastCallSnapshot();
});

test('get list of merge requests', async () => {
  await service.project.listMergeRequests('id id id');
  expect(got.get).toHaveBeenCalled();
  expect(got.get).toMatchLastCallSnapshot();
});

test('get list of closed merge requests', async () => {
  await service.project.listMergeRequests('id id id', { state: 'closed' });
  expect(got.get).toHaveBeenCalled();
  expect(got.get).toMatchLastCallSnapshot();
});

test('get list of branches', async () => {
  await service.project.listBranches('blubb');
  expect(got.get).toHaveBeenCalled();
  expect(got.get).toMatchLastCallSnapshot();
});
