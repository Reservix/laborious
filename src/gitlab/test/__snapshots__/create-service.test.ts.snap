// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a merge request: toMatchLastCallSnapshot 1`] = `
Array [
  "https://gitlab.example.com/api/v4/projects/12346/merge_requests",
  Object {
    "body": Object {
      "id": 12346,
      "source_branch": "feature/123",
      "target_branch": "master",
      "title": "this is a merge request",
    },
    "headers": Object {
      "accept": "application/json",
      "private-token": "T0kn",
      "user-agent": "Testing laborious",
    },
    "json": true,
  },
]
`;

exports[`get a project: toMatchLastCallSnapshot 1`] = `
Array [
  "https://gitlab.example.com/api/v4/projects/ids%20can%20also%20be%20strings",
  Object {
    "headers": Object {
      "accept": "application/json",
      "private-token": "T0kn",
      "user-agent": "Testing laborious",
    },
    "json": true,
  },
]
`;

exports[`get list of branches: toMatchLastCallSnapshot 1`] = `
Array [
  "https://gitlab.example.com/api/v4/projects/blubb/repository/branches",
  Object {
    "headers": Object {
      "accept": "application/json",
      "private-token": "T0kn",
      "user-agent": "Testing laborious",
    },
    "json": true,
  },
]
`;

exports[`get list of closed merge requests: toMatchLastCallSnapshot 1`] = `
Array [
  "https://gitlab.example.com/api/v4/projects/id%20id%20id/merge_requests",
  Object {
    "headers": Object {
      "accept": "application/json",
      "private-token": "T0kn",
      "user-agent": "Testing laborious",
    },
    "json": true,
    "query": Object {
      "state": "closed",
    },
  },
]
`;

exports[`get list of merge requests: toMatchLastCallSnapshot 1`] = `
Array [
  "https://gitlab.example.com/api/v4/projects/id%20id%20id/merge_requests",
  Object {
    "headers": Object {
      "accept": "application/json",
      "private-token": "T0kn",
      "user-agent": "Testing laborious",
    },
    "json": true,
    "query": Object {
      "state": "opened",
    },
  },
]
`;

exports[`get version: toMatchLastCallSnapshot 1`] = `
Array [
  "https://gitlab.example.com/api/v4/version",
  Object {
    "headers": Object {
      "accept": "application/json",
      "private-token": "T0kn",
      "user-agent": "Testing laborious",
    },
    "json": true,
  },
]
`;
